File: pkg/dev_compiler/test/codegen_expected/corelib/string_from_list_test.js
dart_library.library('corelib/string_from_list_test', null, /* Imports */[
  'dart_sdk',
  'expect'
], function load__string_from_list_test(exports, dart_sdk, expect) {
  'use strict';
  const core = dart_sdk.core;
  const _interceptors = dart_sdk._interceptors;
  const dart = dart_sdk.dart;
  const dartx = dart_sdk.dartx;
  const expect$ = expect.expect;
  const string_from_list_test = Object.create(null);
  let ListOfint = () => (ListOfint = dart.constFn(core.List$(core.int)))();
  let JSArrayOfint = () => (JSArrayOfint = dart.constFn(_interceptors.JSArray$(core.int)))();
  let IterableOfint = () => (IterableOfint = dart.constFn(core.Iterable$(core.int)))();
  let intToint = () => (intToint = dart.constFn(dart.definiteFunctionType(core.int, [core.int])))();
  let intTobool = () => (intTobool = dart.constFn(dart.definiteFunctionType(core.bool, [core.int])))();
  let VoidTovoid = () => (VoidTovoid = dart.constFn(dart.definiteFunctionType(dart.void, [])))();
  let const$;
  let const$0;
  let const$1;
  string_from_list_test.main = function() {
    expect$.Expect.equals("", core.String.fromCharCodes(ListOfint().new(0)));
    expect$.Expect.equals("", core.String.fromCharCodes(JSArrayOfint().of([])));
    expect$.Expect.equals("", core.String.fromCharCodes(const$ || (const$ = dart.constList([], core.int))));
    expect$.Expect.equals("AB", core.String.fromCharCodes(JSArrayOfint().of([65, 66])));
    expect$.Expect.equals("AB", core.String.fromCharCodes(const$0 || (const$0 = dart.constList([65, 66], core.int))));
    expect$.Expect.equals("Ærø", core.String.fromCharCodes(const$1 || (const$1 = dart.constList([198, 114, 248], core.int))));
    expect$.Expect.equals("ሴ", core.String.fromCharCodes(JSArrayOfint().of([4660])));
    expect$.Expect.equals("𒍅*", core.String.fromCharCodes(JSArrayOfint().of([74565, 42])));
    expect$.Expect.equals("", core.String.fromCharCodes(ListOfint().new()));
    {
      let a = core.List.new();
      a[dartx.add](65);
      a[dartx.add](66);
      expect$.Expect.equals("AB", core.String.fromCharCodes(IterableOfint()._check(a)));
    }
    for (let len of JSArrayOfint().of([499, 500, 501, 999, 100000])) {
      let list = ListOfint().new(len);
      for (let i = 0; i < dart.notNull(len); i++) {
        list[dartx.set](i, 65 + i[dartx['%']](26));
      }
      for (let i = dart.notNull(len) - 9; i < dart.notNull(len); i++) {
        list[dartx.set](i, 48 + (dart.notNull(len) - i));
      }
      let long = core.String.fromCharCodes(list);
      expect$.Expect.isTrue(long[dartx.startsWith]('ABCDE'));
      expect$.Expect.isTrue(long[dartx.endsWith]('987654321'));
      let middle = (dart.notNull(len) / 2)[dartx.truncate]();
      middle = middle - middle[dartx['%']](26);
      expect$.Expect.equals('XYZABC', long[dartx.substring](middle - 3, middle + 3));
      expect$.Expect.equals(len, long[dartx.length]);
    }
    expect$.Expect.equals("CBA", core.String.fromCharCodes(JSArrayOfint().of([65, 66, 67])[dartx.reversed]));
    expect$.Expect.equals("BCD", core.String.fromCharCodes(JSArrayOfint().of([65, 66, 67])[dartx.map](core.int)(dart.fn(x => dart.notNull(x) + 1, intToint()))));
    expect$.Expect.equals("AC", core.String.fromCharCodes(JSArrayOfint().of([65, 66, 67])[dartx.where](dart.fn(x => x[dartx.isOdd], intTobool()))));
    expect$.Expect.equals("CE", core.String.fromCharCodes(JSArrayOfint().of([65, 66, 67])[dartx.where](dart.fn(x => x[dartx.isOdd], intTobool()))[dartx.map](core.int)(dart.fn(x => dart.notNull(x) + 2, intToint()))));
    expect$.Expect.equals("ABC", core.String.fromCharCodes(IterableOfint().generate(3, dart.fn(x => 65 + dart.notNull(x), intToint()))));
    expect$.Expect.equals("ABC", core.String.fromCharCodes("ABC"[dartx.codeUnits]));
    expect$.Expect.equals("BCD", core.String.fromCharCodes("ABC"[dartx.codeUnits][dartx.map](core.int)(dart.fn(x => dart.notNull(x) + 1, intToint()))));
    expect$.Expect.equals("BCD", core.String.fromCharCodes("ABC"[dartx.runes].map(core.int)(dart.fn(x => dart.notNull(x) + 1, intToint()))));
    let nonBmpCharCodes = JSArrayOfint().of([0, 55314, 56372, 84020, 56372, 55314]);
    let nonBmp = core.String.fromCharCodes(nonBmpCharCodes);
    expect$.Expect.equals(7, nonBmp[dartx.length]);
    expect$.Expect.equals(0, nonBmp[dartx.codeUnitAt](0));
    expect$.Expect.equals(55314, nonBmp[dartx.codeUnitAt](1));
    expect$.Expect.equals(56372, nonBmp[dartx.codeUnitAt](2));
    expect$.Expect.equals(55314, nonBmp[dartx.codeUnitAt](3));
    expect$.Expect.equals(56372, nonBmp[dartx.codeUnitAt](4));
    expect$.Expect.equals(56372, nonBmp[dartx.codeUnitAt](5));
    expect$.Expect.equals(55314, nonBmp[dartx.codeUnitAt](6));
    let reversedNonBmp = core.String.fromCharCodes(nonBmpCharCodes[dartx.reversed]);
    expect$.Expect.equals(7, reversedNonBmp[dartx.length]);
    expect$.Expect.equals(0, reversedNonBmp[dartx.codeUnitAt](6));
    expect$.Expect.equals(55314, reversedNonBmp[dartx.codeUnitAt](5));
    expect$.Expect.equals(56372, reversedNonBmp[dartx.codeUnitAt](4));
    expect$.Expect.equals(56372, reversedNonBmp[dartx.codeUnitAt](3));
    expect$.Expect.equals(55314, reversedNonBmp[dartx.codeUnitAt](2));
    expect$.Expect.equals(56372, reversedNonBmp[dartx.codeUnitAt](1));
    expect$.Expect.equals(55314, reversedNonBmp[dartx.codeUnitAt](0));
    expect$.Expect.equals(nonBmp, core.String.fromCharCodes(nonBmp[dartx.codeUnits]));
    expect$.Expect.equals(nonBmp, core.String.fromCharCodes(nonBmp[dartx.runes]));
  };
  dart.fn(string_from_list_test.main, VoidTovoid());
  // Exports:
  exports.string_from_list_test = string_from_list_test;
});

File: tests/corelib_strong/string_from_list_test.dart
// Copyright (c) 2011, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

import "package:expect/expect.dart";

void main() {
  Expect.equals("", new String.fromCharCodes(new List(0)));
  Expect.equals("", new String.fromCharCodes([]));
  Expect.equals("", new String.fromCharCodes(const []));
  Expect.equals("AB", new String.fromCharCodes([65, 66]));
  Expect.equals("AB", new String.fromCharCodes(const [65, 66]));
  Expect.equals("Ærø", new String.fromCharCodes(const [0xc6, 0x72, 0xf8]));
  Expect.equals("\u{1234}", new String.fromCharCodes([0x1234]));
  Expect.equals("\u{12345}*", new String.fromCharCodes([0x12345, 42]));
  Expect.equals("", new String.fromCharCodes(new List()));
  {
    var a = new List();
    a.add(65);
    a.add(66);
    Expect.equals("AB", new String.fromCharCodes(a));
  }

  // Long list (bug 6919).
  for (int len in [499, 500, 501, 999, 100000]) {
    List<int> list = new List(len);
    for (int i = 0; i < len; i++) {
      list[i] = 65 + (i % 26);
    }
    for (int i = len - 9; i < len; i++) {
      list[i] = 48 + (len - i);
    }
    // We should not throw a stack overflow here.
    String long = new String.fromCharCodes(list);
    // Minimal sanity checking on the string.
    Expect.isTrue(long.startsWith('ABCDE'));
    Expect.isTrue(long.endsWith('987654321'));
    int middle = len ~/ 2;
    middle -= middle % 26;
    Expect.equals('XYZABC', long.substring(middle - 3, middle + 3));
    Expect.equals(len, long.length);
  }

  // Should work with iterables and non-default-lists (http://dartbug.com/8922)
  Expect.equals("CBA", new String.fromCharCodes([65, 66, 67].reversed));
  Expect.equals(
      "BCD", new String.fromCharCodes([65, 66, 67].map((x) => x + 1)));
  Expect.equals(
      "AC", new String.fromCharCodes([0x41, 0x42, 0x43].where((x) => x.isOdd)));
  Expect.equals(
      "CE",
      new String.fromCharCodes(
          [0x41, 0x42, 0x43].where((x) => x.isOdd).map((x) => x + 2)));
  Expect.equals(
      "ABC", new String.fromCharCodes(new Iterable.generate(3, (x) => 65 + x)));
  Expect.equals("ABC", new String.fromCharCodes("ABC".codeUnits));
  Expect.equals(
      "BCD", new String.fromCharCodes("ABC".codeUnits.map((x) => x + 1)));
  Expect.equals("BCD", new String.fromCharCodes("ABC".runes.map((x) => x + 1)));

  var nonBmpCharCodes = [0, 0xD812, 0xDC34, 0x14834, 0xDC34, 0xD812];
  var nonBmp = new String.fromCharCodes(nonBmpCharCodes);
  Expect.equals(7, nonBmp.length);
  Expect.equals(0, nonBmp.codeUnitAt(0));
  Expect.equals(0xD812, nonBmp.codeUnitAt(1)); // Separated surrogate pair
  Expect.equals(0xDC34, nonBmp.codeUnitAt(2));
  Expect.equals(0xD812, nonBmp.codeUnitAt(3)); // Single non-BMP code point.
  Expect.equals(0xDC34, nonBmp.codeUnitAt(4));
  Expect.equals(0xDC34, nonBmp.codeUnitAt(5)); // Unmatched surrogate.
  Expect.equals(0xD812, nonBmp.codeUnitAt(6)); // Unmatched surrogate.

  var reversedNonBmp = new String.fromCharCodes(nonBmpCharCodes.reversed);
  Expect.equals(7, reversedNonBmp.length);
  Expect.equals(0, reversedNonBmp.codeUnitAt(6));
  Expect.equals(0xD812, reversedNonBmp.codeUnitAt(5));
  Expect.equals(0xDC34, reversedNonBmp.codeUnitAt(4));
  Expect.equals(0xDC34, reversedNonBmp.codeUnitAt(3));
  Expect.equals(0xD812, reversedNonBmp.codeUnitAt(2));
  Expect.equals(0xDC34, reversedNonBmp.codeUnitAt(1));
  Expect.equals(0xD812, reversedNonBmp.codeUnitAt(0));

  Expect.equals(nonBmp, new String.fromCharCodes(nonBmp.codeUnits));
  Expect.equals(nonBmp, new String.fromCharCodes(nonBmp.runes));
}
